商品出品機能

murakami-m opened the pull request:
# What
商品出品機能及びitemモデルのテストコードの記述
# Whay
商品出品機能の実装のため

以下GyazoURL
ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/408f5991ad069f58ef42b0dda1984b0a
価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/44d24642e750b0489962dfeae96176f9
必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/8d792f251e88b8f57fdb2b4f6761224d
入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/fd33b898ae53a69353dbbcc6132234fa
ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/3d5568a83d134dd6667b3b6cedb4d446
テスト結果の画像
https://gyazo.com/f81ebbe1773cd64aece80288badaf39f

murakami-m committed:
itemモデルとitemsテーブルの作成、newアクションとcreateアクションの定義、バリデーションの記述

murakami-m committed:
Active Storageの導入

murakami-m committed:
ActiveHashの導入

murakami-m committed:
販売価格のバリデーションを追記

murakami-m committed:
各モデルのクラス名の修正、ビューファイルの修正

murakami-m committed:
価格入力後の手数料と販売利益の表示を実装

murakami-m committed:
ログインの有無でページ移動を変更、その他修正

murakami-m committed:
itemモデルの単体テストコードの記述

Saka-toku commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,31 @@
+    if @item.save
+      redirect_to root_path
+    else
+      @items = Item.includes(:user)
```
`@items = Item.includes(:user)`
上記で定義した@itemsはnew.htmlで使用していないかと存じます。
そのため上記の記述は削除をお願いいたします！

Saka-toku commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,31 @@
+  end
+
+  def move_to_index
+    redirect_to user_session_path unless user_signed_in?
```
`    redirect_to user_session_path unless user_signed_in?`
上記の記述で問題なくログイン画面に遷移しているかと存じます。
ただリクエスト先が「settion#create」に対するpathなっているため、newアクションに対してのpathを記述しましょう！

もしくはdeviseが用意している「authenticate_user!メソッド」を使用しても良いかと存じます。
上記のメソッドを使用することで、記述量を減らすることが可能です。

Saka-toku commented on app/javascript/commission.js:
```diff
@@ -0,0 +1,9 @@
+      const addTaxDom = document.getElementById("add-tax-price");
+      addTaxDom.innerHTML = Math.round(price.value * 0.1 );
+      const addPofitDom = document.getElementById("profit");
+      addPofitDom.innerHTML = Math.round(price.value - Math.round(price.value * 0.1 ))
```
現在の記述ですと小数点切り捨てになっていないかと存じます。
こちらのファイルにある「Math.round」を小数点以下を切り捨てるメソッドに変更をお願いいたします！

【理由】
以下の実装条件を満たすため：
・販売手数料と販売利益は、小数点以下を切り捨てて表示すること。

Saka-toku commented on app/models/item.rb:
```diff
@@ -0,0 +1,23 @@
+    validates :prefecture_id, numericality: { other_than: 1 }
+    validates :days_to_ship_id, numericality: { other_than: 1 }
+    validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999 },
+                      format: { with: /\A[0-9]+\z/ }
```
`validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999 }, format: { with: /\A[0-9]+\z/ }`

上記のformat: { with: /\A[0-9]+\z/ }は削除をお願いいたします。
その代わりの「整数のみ」というバリデーションが必要になるため「整数のみ rails」などと検索いただけますと幸いです。

「理由」
integer型では正規表現が使用できないためです。

Saka-toku commented on app/views/layouts/application.html.erb:
```diff
@@ -5,7 +5,7 @@
   <%= csrf_meta_tags %>
   <%= csp_meta_tag %>
-  <script type="text/javascript" src="https://js.pay.jp/v1/"></script>
+  <script type="text/javascript" src="commission.js"></script>
```
`<script type="text/javascript" src="commission.js"></script>`
上記の記述では購入機能等でエラーになる恐れがあります。
そのため以下の記述に戻すようお願いいたします。
```
  <script type="text/javascript" src="https://js.pay.jp/v1/"></script>
```

Saka-toku commented on db/migrate/20220106060606_create_items.rb:
```diff
@@ -0,0 +1,16 @@
+      t.integer :prefecture_id,      null: false
+      t.integer :days_to_ship_id,    null: false
+      t.integer :price,              null: false
+      t.integer :user_id,            null: false
```
`      t.integer :user_id,            null: false`
usersテーブルの外部キーカラムが適切な形になっていないように見受けられます。

user情報をusersテーブルの外部キーとして扱えるよう、過去アプリを参考に修正をお願いいたします。
また今回はテーブルに関する修正のため「rails db:rollback」をお忘れないようお気をつけください。

Saka-toku reviewed (commented):
レビュー依頼、動画のご提出ありがとうございます！
テストも適切な内容で実装されていて、実装条件をしっかりとご確認されている様子が窺えました！

コメントいたしましたのでご確認をお願いいたします！

murakami-m committed:
商品出品機能における記述の修正

itemsコントローラーの修正、販売手数料・販売利益を小数点以下切り捨てで表示、priceのバリデーションの修正、application.html.erbの修正、マイグレーションファイル及びitemテーブルの修正(ロールバック済)

murakami-m commented:
ご確認ありがとうございました。
ご指摘いただいた箇所を修正致しましたので、再度ご確認の程よろしくお願いいたします。
修正内容が多かったため、Descriptionに詳細を記述しました。

Saka-toku reviewed (commented):
レビュー依頼ありがとうございます。
修正を確認できました！
商品出品の実装ですがLGTMとさせていただきます！
ユーザー機能に続き、今回も大きな実装でしたが無事に乗り越えられましたね！
お疲れさまでした！！

マージ後、ローカルのmasterブランチでプルをお願いいたします。
その後masterブランチから新たなブランチを作成し、次の実装に取り組んでいただけますと幸いです。
次の実装も頑張っていきましょう！
